ssl/timer_queue.c

```diff
@@ -0,0 +1,114 @@
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <stdlib.h>
+#include "internal/timer_queue.h"
+#include "ssl_local.h"  /* for ssl_get_current_time() */
+
+
+static int event_time_compare(const OSSL_TIMER_EVENT *a,
+                              const OSSL_TIMER_EVENT *b)
+{
+    return ossl_time_compare(a->when, b->when);
+}
+
+OSSL_TIMER_QUEUE *ossl_timer_queue_new(void)
+{
+    return ossl_pqueue_OSSL_TIMER_EVENT_new(&event_time_compare);
+}
+
+void ossl_timer_queue_free(OSSL_TIMER_QUEUE *tq)
+{
+    OSSL_TIMER_EVENT *e;
+
+    while ((e = ossl_pqueue_OSSL_TIMER_EVENT_pop(tq)) != NULL)
+        if (e->early_delete != NULL)
+            e->early_delete(tq, e->arg);
+    ossl_pqueue_OSSL_TIMER_EVENT_free(tq);
+}
+
+OSSL_TIMER_EVENT *ossl_timer_queue_new_event(
+        OSSL_TIME when,
+        void (*trigger)(OSSL_TIMER_QUEUE *, void *arg),
+        void (*early_delete)(OSSL_TIMER_QUEUE *, void *arg),
+        void *arg)
+{
+    OSSL_TIMER_EVENT *e = OPENSSL_malloc(sizeof(*e));
+
+    if (e != NULL)
+        ossl_timer_queue_event(e, when, trigger, early_delete, arg);
+    return e;
+}
+
+void ossl_timer_queue_free_event(OSSL_TIMER_EVENT *event)
+{
+    OPENSSL_free(event);
+}
+
+int ossl_timer_queue_schedule(OSSL_TIMER_QUEUE *tq, OSSL_TIMER_EVENT *event)
+{
+    return ossl_pqueue_OSSL_TIMER_EVENT_push(tq, event, &event->ref);
+}
+
+int ossl_timer_queue_delete(OSSL_TIMER_QUEUE *tq, OSSL_TIMER_EVENT *event)
+{
+    if (!ossl_pqueue_OSSL_TIMER_EVENT_remove(tq, event->ref))
+        return 0;
+    if (event->early_delete != NULL)
+           event->early_delete(tq, event->arg);
+    return 1;
+}
+
+OSSL_TIME ossl_timer_queue_time_until(OSSL_TIMER_EVENT *event)
+{
+    if (event == NULL)
+        return OSSL_TIME_NEVER;
+    return ossl_time_subtract(event->when, ossl_time_now());
+}
+
+OSSL_TIME ossl_timer_queue_time_until_next(const OSSL_TIMER_QUEUE *tq)
+{
+    if (tq == NULL)
+        return OSSL_TIME_NEVER;
+    return ossl_timer_queue_time_until(ossl_pqueue_OSSL_TIMER_EVENT_peek(tq));
+}
+
+int ossl_timer_queue_postpone(OSSL_TIMER_QUEUE *tq, OSSL_TIMER_EVENT *event,
+                              OSSL_TIME how_long)
+{
+    return ossl_timer_queue_postpone_until(tq, event, ossl_time_add(event->when,
+                                                                    how_long));
+}
+
+int ossl_timer_queue_postpone_until(OSSL_TIMER_QUEUE *tq,
+                                    OSSL_TIMER_EVENT *event,
+                                    OSSL_TIME when)
+{
+    if (ossl_pqueue_OSSL_TIMER_EVENT_remove(tq, event->ref)) {
+        event->when = when;
+        return ossl_timer_queue_schedule(tq, event);
+    }
+    return 0;
+}
+
+int ossl_timer_queue_execute(OSSL_TIMER_QUEUE *tq)
+{
+    OSSL_TIME now = ossl_time_now();
+    OSSL_TIMER_EVENT *e;
+    int n = 0;
+
+    while (ossl_time_compare(ossl_pqueue_OSSL_TIMER_EVENT_peek(tq)->when,
```

Comment:
Why is this a while loop?

Reply:
It is meant to handle all events that are due.

It would be easy enough to only process one event per call.

I'm not sure which is better.